File: Contents/code/haskel/chapter4/apply_twice.hs
-- A simple pure function
addOne :: Int -> Int
addOne x = x + 1

multiplyByTwo :: Int -> Int
multiplyByTwo x = x * 2

-- HIGHER-ORDER FUNCTION
-- Takes a function (Int -> Int) and an Int, returns an Int
applyTwice :: (Int -> Int) -> Int -> Int
applyTwice f x = f (f x)

main :: IO ()
main = do
  print (applyTwice addOne 5)         -- 7
  print (applyTwice multiplyByTwo 5) -- 20

File: Contents/code/python/chapter4/apply_twice.py
from typing import Callable

# A simple pure function
def add_one(x: int) -> int:
    return x + 1

def multiply_by_two(x: int) -> int:
    return x * 2

# HIGHER-ORDER FUNCTION
# Takes a function 'f' and an integer 'x'
def apply_twice(f: Callable[[int], int], x: int) -> int:
    return f(f(x))

if __name__ == "__main__":
    print(apply_twice(add_one, 5))         # (5 + 1) + 1 = 7
    print(apply_twice(multiply_by_two, 5)) # (5 * 2) * 2 = 20
